### PR TITLE
refactor(toml): Pull out the schema

### PR DESCRIPTION
**Pull request by epage**
### What does this PR try to resolve?

On its own, this PR is a net negative for readability / complexity.  It moves all of the serde types related to manifest from `toml/mod.rs` to `toml/schema.rs`, leaving a lot of the functions and some trait implementations back in `toml/mod.rs` (some basic functions that made sense for the type on their own were also moved).

So why do this?  This is an ooch towards having the schema broken out into a separate package (#12801).  To do this, we need to
- Identify what all types need to be put in the package.  This refactor highlights the dependence on `RustVersion` and `PackageIdSpec`
- Highlights what functionality we need to find a new home for

Follow up PRs would
- Find better homes for the logic in `toml/mod.rs`, moving us away from having `impl schema::Type` blocks.
- Pull out a `src/cargo/util_semver` package to own `PartialVersion` (at least) as prep for making a `cargo-util-semver` package
- Move `RustVersion` to `manifest-toml/schema.rs`, deciding what functionality needs to move with the type
- Move or copy `PackageIdSpec` into `manfest-toml/schema.rs`, deciding what functionality remain in `core/` and what moves over
- Move `toml/schema.rs` to `src/cargo/util_schema`
- Actually make `cargo-util-semver` and `cargo-util-manifest-schema` packages

So why do this now? This is a big change!  By being incremental
- Reduce churn for me and others
- Be easier to review
- Collect feedback as we go on the whole plan to avoid more painful changes later

We *can* back this out if needed but the further we go, the more painful it will be.

### CODE REVIEW

**Comment by ehuss** [comment]
Just to confirm, this just moves things, there aren't any other changes of substance?

If so, it sounds good to me. However, I'd like to defer to @weihanglo to see what they would like.

r? @weihanglo

**Review by weihanglo — COMMENTED** [review]
The module name doesn't catch my eye. `toml::schema` looks pretty genernal that we might also want `config.toml` to be inside. I don't think that's the better direction, but is it in the plan in your mind?

File: src/cargo/util/toml/schema.rs
Comment (weihanglo) [inline src/cargo/util/toml/schema.rs:148]: What is the use case of making this public?

File: src/cargo/util/toml/schema.rs
Comment (weihanglo) [inline src/cargo/util/toml/schema.rs:462]: This seems like for emitting a better warning only. Do we want to pub this?

**Review by weihanglo — APPROVED** [review]
While I left two comments about the visibility of seemingly internal only fields, I am going to merge this to unblock stuff.

**Comment by weihanglo** [comment]
> Correct. The most visible changes are for cargo-as-a-library users
> 
>     * Types moved
> 
>     * More types are pub
> 
>     * Fields are pub

Worth noting that, making fields and types `pub` doesn't mean they are stable. There is still no strong stability guarantee for Cargo as a library, but this PR is a path forward that goal.

@bors r+

File: src/cargo/util/toml/schema.rs
Comment (epage) [reply src/cargo/util/toml/schema.rs:148]: That depends on if we want this to be `#[non_exhaustive]` or directly constructable.  For now, I was going with directly constructable.

**Comment by epage** [comment]
Blocked on #12921

**Pull request merged**
